File: p.30/TS/solution.ts
const digitPowers: number[] = Array.from({ length: 10 }, (_, d) => d ** 5);

function sumOfFifthPowersOfDigits(num: number): number {
  let sum = 0;
  let temp = num;

  while (temp > 0) {
    const digit = temp % 10;
    sum += digitPowers[digit];
    temp = Math.floor(temp / 10);
  }

  return sum;
}

function sumOfFifthPowerNumbers(): number {
  let total = 0;

  for (let n = 2; n <= 999_999; n++) {
    if (n === sumOfFifthPowersOfDigits(n)) {
      total += n;
    }
  }

  return total;
}

function main(): void {
  const answer = sumOfFifthPowerNumbers();
  console.log(answer);
}

main();

File: p.30/Kotlin/solution.kt
import kotlin.math.pow

val digitPowers: List<Int> = List(10) {d -> d.toDouble().pow(5).toInt() }

fun isSumOfFifthPowers(num: Int): Boolean {
    var tmp = num
    var sum = 0
    while (tmp > 0) {
        val digit = tmp % 10
        sum += digitPowers[digit]
        tmp /= 10
    }
    return sum == num
}

fun sumOfFifthPowerNumbers(): Long {
    return (2..999_999)
        .filter { isSumOfFifthPowers(it) }
        .map { it.toLong() }
        .sum()
}

fun main() {
    val answer = sumOfFifthPowerNumbers()
    println("$answer")
}
